stan/math/prim/functor/multi_expression.hpp

```diff
@@ -0,0 +1,196 @@
+ * @param exprs exprs that will be used in same kernel.
+ */
+template <typename... T_expressions,
+          require_all_eigen_t<T_expressions...>* = nullptr>
+eigen_expressions_<T_expressions...> eigen_expressions(
+    T_expressions&&... exprs) {
+  return eigen_expressions_<T_expressions...>(
+      std::forward<T_expressions>(exprs)...);
+}
+
+/**
+ * Represents results that will be calculated in same loop.
+ * @tparam T_results types of results
+ */
+template <typename... T_results>
+class eigen_results_ {
+  std::tuple<T_results...> results_;
+
+  /**
+   * Assign expressions to results using linear indexing.
+   * @tparam Linear whether to use linear indexing
+   * @tparam T_expressions types of expressions
+   * @param expressions expressions to assign
+   */
+  template <bool Linear, typename... T_expressions,
+            std::enable_if_t<Linear>* = nullptr>
+  void assign(const eigen_expressions_<T_expressions...>& expressions) {
+    for (size_t i = 0; i < std::get<0>(expressions.exprs_).size(); i++) {
+      index_apply<sizeof...(T_results)>([&](auto... Is) {
+        static_cast<void>(std::initializer_list<int>{
+            (std::get<Is>(results_).coeffRef(i)
+             = std::get<Is>(expressions.exprs_).coeff(i),
+             0)...});
+      });
+    }
+  }
+  /**
+   * Assign expressions to results using 2d indexing.
+   * @tparam Linear whether to use linear indexing
+   * @tparam T_expressions types of expressions
+   * @param expressions expressions to assign
+   */
+  template <bool Linear, typename... T_expressions,
+            std::enable_if_t<!Linear>* = nullptr>
+  void assign(const eigen_expressions_<T_expressions...>& expressions) {
+    constexpr bool is_first_row_major
+        = Eigen::internal::traits<std::decay_t<
+              std::tuple_element_t<0, std::tuple<T_expressions...>>>>::Flags
+          & Eigen::RowMajorBit;
+    size_t outer_dimension = is_first_row_major
+                                 ? std::get<0>(expressions.exprs_).rows()
+                                 : std::get<0>(expressions.exprs_).cols();
+    size_t inner_dimension = is_first_row_major
+                                 ? std::get<0>(expressions.exprs_).cols()
+                                 : std::get<0>(expressions.exprs_).rows();
+    for (size_t i = 0; i < outer_dimension; i++) {
+      for (size_t j = 0; j < inner_dimension; j++) {
+        index_apply<sizeof...(T_results)>([&](auto... Is) {
+          static_cast<void>(std::initializer_list<int>{
+              ((Eigen::internal::traits<std::decay_t<std::tuple_element_t<
+                            Is, std::tuple<T_expressions...>>>>::Flags
+                        & Eigen::RowMajorBit
+                    ? std::get<Is>(results_).coeffRef(j, i)
+                      = std::get<Is>(expressions.exprs_).coeff(j, i)
+                    : std::get<Is>(results_).coeffRef(i, j)
+                      = std::get<Is>(expressions.exprs_).coeff(i, j)),
+               0)...});
```

Comment:
Kind of confused on the row major checking. Is this assuming if the expression is row/col major then the result is also row/col major?

Are there things in Eigen that are row major by default? The only thing i can think of where we would want mixed row/col major support would be something like sparse matrices

(on a side note getting these to work with sparse matrices would be cool because if we used them in function then iterators would also support sparse stuff. but that can happen later)

Reply:
> Is this assuming if the expression is row/col major then the result is also row/col major?

No. It is just choosing whether to go over the values in row or col major order depending on the expression. If one of the (result, expression) is row major and another is col major it does not matter. If both have the same storage order one of these options is much faster (but both would produce the correct result).

> Are there things in Eigen that are row major by default?

Return type of transposing any col major expression comes to mind. There might also be others.